### ISSUE TITLE
Outdated translations can crash individual pages

### ISSUE THREAD

**Issue by jfly** [thewca/worldcubeassociation.org#1259]
So, the /edit/admin page doesn't load for me in locale: es.

Here's what I see on the server logs:

```
F, [2017-02-01T02:45:07.764810 #26469] FATAL -- :
ActionView::Template::Error (missing interpolation argument :comps in "%{comps} %{comp} en %{kms} km y %{days} días." ({:x_competitions=>"9 competitions", :kms=>200, :days=>365} given)):
    35:     <div class="col-sm-9">
    36:       <% nearby_competitions_info = @competition.nearby_competitions(Competition::NEARBY_DAYS_INFO, Competition::NEARBY_DISTANCE_KM_INFO).sort_by(&:start_date) %>
    37:       <% nearby_count = nearby_competitions_info.count %>
    38:       <div class="alert <%= nearby_count < Competition::NEARBY_INFO_COUNT ? "alert-success" : "alert-danger" %>" role="alert"><%= t '.nearby_admin', x_competitions: t('.competitions', count: nearby_count), kms: Competition::NEARBY_DISTANCE_KM_INFO, days: Competition::NEARBY_DAYS_INFO %>
    39:       <button type="button" class="btn" data-toggle="modal" data-target="#competitions-last-year-modal"><%= t '.show' %></button>
    40:       </div>
    41:     </div>
  app/views/competitions/_nearby_competitions.html.erb:38:in `_app_views_competitions__nearby_competitions_html_erb__2076549207874311778_128092340'
  app/views/competitions/_competition_form.html.erb:207:in `block in _app_views_competitions__competition_form_html_erb__2399980640299519409_103775980'
  app/helpers/application_helper.rb:151:in `block in simple_form_for'
  app/helpers/application_helper.rb:150:in `simple_form_for'
  app/helpers/application_helper.rb:168:in `horizontal_simple_form_for'
  app/views/competitions/_competition_form.html.erb:3:in `_app_views_competitions__competition_form_html_erb__2399980640299519409_103775980'
  app/views/competitions/edit.html.erb:18:in `block in _app_views_competitions_edit_html_erb__3953575044078717486_134356900'
  app/views/competitions/_nav.html.erb:193:in `_app_views_competitions__nav_html_erb__2462345723200062896_66947420'   
  app/views/competitions/edit.html.erb:3:in `_app_views_competitions_edit_html_erb__3953575044078717486_134356900'
  app/controllers/competitions_controller.rb:279:in `admin_edit'
  lib/middlewares/warden_user_logger.rb:12:in `call'
  lib/middlewares/fix_accept_header.rb:13:in `call'
```

And if you check https://www.worldcubeassociation.org/server-status, you will see that the competitions > nearby_competitions > nearby_admin key is out of date, which is probably what's causing this error.

It's rather unfortunate that an out of date translation can effectively crash our website. It would be cool if we fell back to English for these cases, or even just inserted an empty string or some warning instead. Any ideas for working around this?

Crazy idea #1: define our own "t" method that catches this error and rerenders in English. If English crashes, that's a-okay, we did our best.

@viroulep said:

This [commit](https://github.com/thewca/worldcubeassociation.org/commit/7a416030a90394359835268be993f1c8d7685a9a) introduced the variables changes in the key, and you can see that the Spanish one has been forgotten!

The crazy idea sounds fun, but this can enforced statically so we can just add a test for this!

@jfly wonders:

That sounds even better! Do you mean "in theory" it can be enforced statically, or is there already a tool for this? Looking through i18n-tasks, I see https://github.com/glebm/i18n-tasks/issues/192, which is still open =(

**Comment by jfly** [comment]
Whether we go the path of writing a test that checks this statically, or a runtime fallback to English, I think this is problematic enough that we should hold of on deploying our i18n changes until we've got a workaround for this. Do you agree, @viroulep?

**Comment by larspetrus** [comment]
On Wed, Feb 1, 2017 at 11:17 AM, Jeremy Fleischman <jeremyfleischman@gmail.com> wrote:
We *do* actually have fallbacks enabled. The problem in this particular case is that this is not a missing translation, it's an out of date translation.

Oh! It's a translation that exists but throws an exception? Interesting...

You can define an exception handler: http://edgeguides.rubyonrails.org/i18n.html#using-different-exception-handlers

**Comment by jfly** [comment]
Nice find! That is **exactly** what we need. Of course, checking this statically would be ideal (and is theoretically possible), but I'm okay with this solution.

I'll let @viroulep decide.

**Comment by larspetrus** [comment]
I assume static analysis would mean you have to wait for all translations to be updated before committing/deploying?

**Comment by jfly** [comment]
Oh, you're totally right. It's not acceptable to have to wait for translations. So, we *definitely* have to implement the fallback.

Static analysis could be nice to help detect mistakes that people make when translating, but we cannot have those errors fail travis (and thereby block merging our code).

**Comment by viroulep** [comment]
Here is a more precise description of the issue and why it happened:

at some point we had key `blah` with content `some %{a} var and %{b} var`.
This key got translated in all languages.
Later we decided to change the content of the key to `some %{c} var`.

This change means that the translated version will throw an exception because the app will not be providing `a` nor `b` anymore.
So this change must come with either:
1. a removal of the translated keys in all translations
2. an update of the translated keys

For this particular example, it was a trivial change so I went for option 2, but unfortunately forgot to change the Spanish one, leading to this issue.

Some general thoughts now:
 - @jonatanklosko's app enforces a perfect match between the sets of variables in the original string and in the translated one.
 - This is totally possible to check this statically in specs.
 - The same issue can be solved by implementing the custom exception handler

The big advantage of the exception handler option is that programmers don't need to touch locales (except English).
